python/src/pyarrow.rs

```diff
@@ -0,0 +1,205 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+use libc::uintptr_t;
+use pyo3::exceptions::PyNotImplementedError;
+use pyo3::prelude::*;
+use pyo3::types::PyList;
+
+use datafusion::arrow::array::{make_array_from_raw, ArrayRef};
+use datafusion::arrow::datatypes::{DataType, Field, Schema};
+use datafusion::arrow::ffi;
+use datafusion::arrow::ffi::FFI_ArrowSchema;
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::scalar::ScalarValue;
+
+use crate::errors::DataFusionError;
+
+pub trait PyArrowConvert: Sized {
+    fn from_pyarrow(value: &PyAny) -> PyResult<Self>;
+    fn to_pyarrow(&self, py: Python) -> PyResult<PyObject>;
+}
+
+impl PyArrowConvert for DataType {
```

Comment:
I'm going to try adding this module to `arrow-rs` as an optional one, so we can implement the `PyO3` conversion traits, like `FromPyObject` directly for the arrow types. 
This will further reduce the required conversion boilerplate in the python bindinds. 

For example we should be able to write
```rust
fn create_dataframe(
        &mut self,
        partitions: Vec<Vec<RecordBatch>>,
    ) -> PyResult<PyDataFrame> {
        // partitions are going to be converted by PyO3 automatically
```

instead of

```rust
fn create_dataframe(
        &mut self,
        partitions: Vec<Vec<&PyAny>>,
    ) -> PyResult<PyDataFrame> {
        let partitions: Vec<Vec<RecordBatch>> = partitions
            .into_iter()
            .map(|batches| {
                batches
                    .into_iter()
                    .map(RecordBatch::from_pyarrow)
                    .collect::<PyResult<_>>()
            })
            .collect::<PyResult<_>>()?;
```

Reply:
I admit part of the reason I avoided something like that at the time was to avoid canibalizing pyarrow, since it makes it appealing from there to have arrow-rs exposed as a Python library, which imo would add confusion to the Python ecosystem (having two official Python libraries, one from each implementation).

OTOH, polars, and likely others, would benefit from such a pyo3 library, as anyone building bindings on top of arrow-rs could re-use that code.

Reply:
@jorgecarleitao created a prototype, see the referenced PRs below. I guess the code should be portable to arrow2 as well.